### PR TITLE
RISC-V: Support CFG_PAN and some fixing

### PR DESCRIPTION
**Pull request by gagachang**
The series add support `CFG_PAN` for RISC-V architecture.
Some fixing on thread RPC and bit-width of CSR status are also submitted.

### CODE REVIEW

**Comment by gagachang** [comment]
Hi @maroueneboubakri 

I modified some of you code.
Please let me know if you have doubt, thanks!

**Comment by maroueneboubakri** [comment]
@gagachang for https://github.com/OP-TEE/optee_os/pull/6184/commits/6f70b3d36ccf9124f5237d7c40bc8b99ee01d797 why are you calling it wrong stack usage and proposing a fix for stack overflow ?  I'm using `s0` to hold a value while you propose to use it to hold the frame pointer and you save it in addition to the stack for your stack unwinding, I do agree without but the commit description needs to be reconsidered.
Thank you for the code regarding the status CSR.

**Comment by maroueneboubakri** [comment]
For commit https://github.com/OP-TEE/optee_os/pull/6184/commits/81e2eb8144f4e9a678bbe467a3e6465ab69fa966 I don't see the motivation behind saving the callee saved registers in the trap handler.

**Comment by maroueneboubakri** [comment]
Commit https://github.com/OP-TEE/optee_os/pull/6184/commits/0f7ba637e318ddeb3cae0da4cbb817fa4182f1df 

`Tested-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

**Comment by gagachang** [comment]
> @gagachang for [6f70b3d](https://github.com/OP-TEE/optee_os/commit/6f70b3d36ccf9124f5237d7c40bc8b99ee01d797) why are you calling it wrong stack usage and proposing a fix for stack overflow ? 

Because in original implementation, the `sp` is deceased by 16 bytes.
But there are three registers saved onto the stack, which should occupy 3*8 = 24 bytes in RV64.

```
 /* Use stack for temporary storage */
addi	sp, sp, -16

...

/* Save return address xSTATUS and pointer to rv */
STR	a0, REGOFF(0)(sp)
STR	a1, REGOFF(1)(sp)
STR	ra, REGOFF(2)(sp)
```

**Comment by maroueneboubakri** [comment]
> > @gagachang for [6f70b3d](https://github.com/OP-TEE/optee_os/commit/6f70b3d36ccf9124f5237d7c40bc8b99ee01d797) why are you calling it wrong stack usage and proposing a fix for stack overflow ?
> 
> Because in original implementation, the `sp` is deceased by 16 bytes. But there are three registers saved onto the stack, which should occupy 3*8 = 24 bytes in RV64.
> 
> ```
>  /* Use stack for temporary storage */
> addi	sp, sp, -16
> 
> ...
> 
> /* Save return address xSTATUS and pointer to rv */
> STR	a0, REGOFF(0)(sp)
> STR	a1, REGOFF(1)(sp)
> STR	ra, REGOFF(2)(sp)
> ```

Correct !

**Comment by gagachang** [comment]
> For commit [81e2eb8](https://github.com/OP-TEE/optee_os/commit/81e2eb8144f4e9a678bbe467a3e6465ab69fa966) I don't see the motivation behind saving the callee saved registers in the trap handler.

Yes you are right.
I thought asynchronous notification triggers OP-TEE context switch, but I think I am wrong after more study.
I will remove this commit.

**Comment by maroueneboubakri** [comment]
For commit https://github.com/OP-TEE/optee_os/pull/6184/commits/ef81c6a36206604d248e2b6f4f75a8422a1f5869 

`Reviewed-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

**Comment by gagachang** [comment]
Thank you @maroueneboubakri 
There is one remaining commit to be reviewed.

**Comment by maroueneboubakri** [comment]
> Dear @maroueneboubakri Kindly request for last commit [d651cf9](https://github.com/OP-TEE/optee_os/commit/d651cf9a2b051b36b11020e92c802e1a51abb29a) to be reviewed. Thanks!

@gagachang sure but please addrress the comments above

**Comment by gagachang** [comment]
@maroueneboubakri 

Sorry but I don't see any comments related to commit https://github.com/OP-TEE/optee_os/pull/6184/commits/d651cf9a2b051b36b11020e92c802e1a51abb29a

**Comment by maroueneboubakri** [comment]
For commit https://github.com/OP-TEE/optee_os/commit/d651cf9a2b051b36b11020e92c802e1a51abb29a with my comment handled

`Reviewed-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

**Comment by gagachang** [comment]
@maroueneboubakri 
Comments are addressed, thanks!

**Comment by gagachang** [comment]
> @gagachang stills one comment above about the use of `ra` as trampoline to restore `xSTATUS`, can you check it ?

Sure, but I don't see any comments when I check the PR.
Could you give me more detail?

**Comment by maroueneboubakri** [comment]
> > @gagachang stills one comment above about the use of `ra` as trampoline to restore `xSTATUS`, can you check it ?
> 
> Sure, but I don't see any comments when I check the PR. Could you give me more detail?

Strange, here it is:

image

**Comment by jforissier** [comment]
> > > @gagachang stills one comment above about the use of `ra` as trampoline to restore `xSTATUS`, can you check it ?
> > 
> > 
> > Sure, but I don't see any comments when I check the PR. Could you give me more detail?
> 
> Strange, here it is:
> 
> image

"Pending" means you have not submitted the review yet I think :wink:

File: core/arch/riscv/kernel/thread_rv.S [hunk 1]
```diff
@@ -44,14 +45,12 @@
 	la	gp, __global_pointer$
 .option pop
 .endif
-	store_xregs sp, THREAD_TRAP_REG_T3, REG_T3, REG_T6
 	store_xregs sp, THREAD_TRAP_REG_T0, REG_T0, REG_T2
+	store_xregs sp, THREAD_TRAP_REG_S0, REG_S0, REG_S1
 	store_xregs sp, THREAD_TRAP_REG_A0, REG_A0, REG_A7
+	store_xregs sp, THREAD_TRAP_REG_S2, REG_S2, REG_S11
+	store_xregs sp, THREAD_TRAP_REG_T3, REG_T3, REG_T6
```
[the remark this answers is not in the document]
Comment (gagachang) [reply core/arch/riscv/kernel/thread_rv.S]: That commit has already been removed

File: core/arch/riscv/kernel/thread_rv.S [hunk 2]
```diff
@@ -376,9 +381,14 @@ FUNC thread_rpc , :
 	sw	a2, 8(a4)
 	sw	a3, 12(a4)
 
-	/* Pop return address from stack */
-	LDR	ra, REGOFF(2)(sp)
+	/* Pop saved XSTATUS from stack */
+	LDR	ra, REGOFF(1)(sp)
+	csrw	CSR_XSTATUS, ra
```
[the remark this answers is not in the document]
Comment (gagachang) [reply core/arch/riscv/kernel/thread_rv.S:386]: Maybe we can use `s0` instead of `ra` here

[the remark this answers is not in the document]
Comment (gagachang) [reply core/arch/riscv/kernel/thread_rv.S]: Addressed by using `s0`

**Comment by gagachang** [comment]
@maroueneboubakri 

I notice that we should use the return value of `set_field_u64()`, please confirm it.

```C
- set_field_u64(status, CSR_XSTATUS_SPP, PRV_U);
+ status = set_field_u64(status, CSR_XSTATUS_SPP, PRV_U);
```

**Comment by maroueneboubakri** [comment]
> @maroueneboubakri
> 
> I notice that we should use the return value of `set_field_u64()`, please confirm it.
> 
> ```c
> - set_field_u64(status, CSR_XSTATUS_SPP, PRV_U);
> + status = set_field_u64(status, CSR_XSTATUS_SPP, PRV_U);
> ```

yes I confirm.

**Comment by gagachang** [comment]
Hello @jforissier 

CI shows the following error and need your help
```
zstd: error 2[5](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:5) : Write error : No space left on device (cannot write compressed block) 
/usr/bin/tar: cache.tzst: Cannot write: Broken pipe
/usr/bin/tar: Child returned status 25
/usr/bin/tar: Error is not recoverable: exiting now
Warning: Failed to save: "/usr/bin/tar" failed with error: The process '/usr/bin/tar' failed with exit code 2
Error: Unable to process command 'warning' successfully.
Error: No space left on device : '/home/runner/runners/2.308.0/_diag/pages/ad0aa8d7-3f81-4fec-8704-8e8f4a934afb_[6](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:6)9e[7](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:7)4230-6ee[8](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:8)-54a0-a215-afc07c[9](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:9)0c202_2.log'
```

**Comment by jforissier** [comment]
> Hello @jforissier
> 
> CI shows the following error and need your help
> 
> ```
> zstd: error 2[5](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:5) : Write error : No space left on device (cannot write compressed block) 
> /usr/bin/tar: cache.tzst: Cannot write: Broken pipe
> /usr/bin/tar: Child returned status 25
> /usr/bin/tar: Error is not recoverable: exiting now
> Warning: Failed to save: "/usr/bin/tar" failed with error: The process '/usr/bin/tar' failed with exit code 2
> Error: Unable to process command 'warning' successfully.
> Error: No space left on device : '/home/runner/runners/2.308.0/_diag/pages/ad0aa8d7-3f81-4fec-8704-8e8f4a934afb_[6](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:6)9e[7](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:7)4230-6ee[8](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:8)-54a0-a215-afc07c[9](https://github.com/OP-TEE/optee_os/actions/runs/5976731020/job/16215203936?pr=6184#step:9:9)0c202_2.log'
> ```

There is not enough space in the filesystem of the CI runner. The error occurs as the CI job is creating an archive of the compiler cache. That's annoying, and I am not sure how we can use less dick space at this point. Perhaps by reducing the maximum size of the ccache cache (`ccache -M size`). I will investigate.

**Comment by jforissier** [comment]
Commit "core: riscv: Support Privileged Access Never by CSR status.SUM bit" still needs a Reviewed-by or an Acked-by, @maroueneboubakri perhaps?

**Comment by maroueneboubakri** [comment]
Commit "core: riscv: Support Privileged Access Never by CSR status.SUM bit"
`Reviewed-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

**Pull request merged**